### PR TITLE
Simplified the StartsActivity by reducing the number of parameters through POJO class

### PR DESCRIPTION
**Pull request by email2vimalraj**
Simplified the StartsActivity by reducing the number of parameters through POJO class

## Change list

Currently the number of `startActivity` methods are unnecessary which can be solved by one single method with single parameter. So introduced the simple POJO class which in future can be expanded with any number of parameters.
 
## Types of changes

What types of changes are you proposing/introducing to Java client?

- [ ] No changes in production code.
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Details

The sample usage of new implementation is as follows:

```java
Activity activity = new Activity();
activity.setAppPackage("com.foo");
activity.setAppActivity(".bar");
driver.startActivity(activity);
```

### CODE REVIEW

**Comment by email2vimalraj** [comment]
Could someone please explain why the CI build has failed? 
I'm unable to run the build  in my machine. Getting compilation error.
Please have a look at the exception that I'm getting: https://gist.github.com/email2vimalraj/171ecdaa966db76e047e0cc7e2942481

**Comment by email2vimalraj** [comment]
@SrinivasanTarget : Kindly help me what I'm supposed to fix in the build error.

**Comment by saikrishna321** [comment]
@email2vimalraj you have styling issues. Please fix them 

https://travis-ci.org/appium/java-client/builds/204062313#L1987

you can ignore the warnings

**Comment by SrinivasanTarget** [comment]
@email2vimalraj Hey buddy sorry just checked your note. please fix the errors to review.

File: src/main/java/io/appium/java_client/android/StartsActivity.java [hunk 1]
```diff
@@ -26,93 +26,26 @@
...
+     * <pre>
+     *     {@code
+     *     Activity activity = new Activity();
+     *     activity.setAppPackage("com.foo");
+     *     activity.setAppActivity(".bar");
+     *     driver.startActivity(activity);
+     *     }
+     * </pre>
      *
-     * @param appPackage      The package containing the activity. [Required]
```
Comment (TikhomirovSergey) [inline src/main/java/io/appium/java_client/android/StartsActivity.java]: @email2vimalraj I'm against this change as it is. Please mark these methods `Deprecated` and with the description to use new ones

File: src/main/java/io/appium/java_client/android/Activity.java [hunk 2]
```diff
@@ -0,0 +1,191 @@
+package io.appium.java_client.android;
+
+/**
+ * This is a simple POJO class to support the {@link StartsActivity}.
+ */
+public class Activity {
```
Comment (TikhomirovSergey) [inline src/main/java/io/appium/java_client/android/Activity.java]: I this this POJO should have 2 mandatory parameteres: package and activity. Please add the constructor with these parameters.

File: src/main/java/io/appium/java_client/android/Activity.java [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+    public String getAppPackage() {
+        return appPackage;
+    }
+
+    /**
+     * Sets the app package value.
+     *
+     * @param appPackage The app package value.
+     */
+    public void setAppPackage(String appPackage) {
```
Comment (SrinivasanTarget) [inline src/main/java/io/appium/java_client/android/Activity.java]: I like the idea behind this but have few things to be discussed here,

Is there a way to remove getters & setters? May be we need to come up with either custom annotation or something like https://github.com/rzwitserloot/lombok.

Comment (email2vimalraj) [inline src/main/java/io/appium/java_client/android/Activity.java]: Read the user review who was using the lombok for more than a year: http://stackoverflow.com/a/12807937/1505987

I would recommend to stay away from lombok, even though I never used it personally. My justification would be obviously a javadoc and everytime we will have to generate the code.

**Comment by email2vimalraj** [comment]
The CI is still pending from long time. Could someone look into it?

File: src/main/java/io/appium/java_client/android/Activity.java (hunk 3 above)
Comment (SrinivasanTarget) [inline src/main/java/io/appium/java_client/android/Activity.java]: I also don't recommend lombock. But i liked their way of using annotations.

**Comment by SrinivasanTarget** [comment]
> The CI is still pending from long time. Could someone look into it?

Have restarted travisCI lets see. Anyways we need to wait for Codacy to complete.

**Comment by email2vimalraj** [comment]
> Have restarted travisCI job lets see. Anyways we need to wait for Codacy to complete.
Seems some issue with Travis, it keeps the job in queue, but not running it up.

**Comment by SrinivasanTarget** [comment]
> Seems some issue with Travis, it keeps the job in queue, but not running it up.

Yes checking

**Comment by email2vimalraj** [comment]
Any updates on CI?

**Comment by email2vimalraj** [comment]
@SrinivasanTarget @TikhomirovSergey : Can any one of you please help on the Travis issue? Seems the travis doesn't like to run my PR :(

**Comment by SrinivasanTarget** [comment]
@email2vimalraj I'll check with @imurchie and see if we could do something.

**Comment by email2vimalraj** [comment]
Seems Travis finally ran my PR and this time build passed. :)

File: docs/The-starting-of-an-Android-app.md [hunk 4]
```diff
@@ -118,7 +119,10 @@ StartsActivity startsActivity = new StartsActivity() {
     }
 };
 
-StartsActivity startsActivity.startActivity("your.package.name", ".ActivityName");
+Activity activity = new Activity();
+activity.setAppPackage("your.package.name");
```
Comment (SrinivasanTarget) [inline docs/The-starting-of-an-Android-app.md]: @email2vimalraj Can you update here too.

Comment (SrinivasanTarget) [inline docs/The-starting-of-an-Android-app.md]: @email2vimalraj ping

**Comment by email2vimalraj** [comment]
@SrinivasanTarget : Fixed, kindly have a look

File: src/main/java/io/appium/java_client/android/Activity.java [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+    private String optionalIntentArguments;
+    private boolean stopApp;
+
+    /**
+     * A constructor with two mandatory parameters.
+     *
+     * @param appPackage The value for the app package.
+     * @param appActivity The value for the app activity.
+     */
+    public Activity(String appPackage, String appActivity) {
```
Comment (TikhomirovSergey) [inline src/main/java/io/appium/java_client/android/Activity.java]: @email2vimalraj Could you provide these checkings:

```java
import static com.google.common.base.Preconditions.checkArgument;
import static org.apache.commons.lang3.StringUtils.isBlank;

public Activity(String appPackage, String appActivity) {
    checkArgument(!isBlank(appPackage), "App package should be defined as not empty or null string");
    checkArgument(!isBlank(appActivity), "App activity should be defined as not empty or null string");
    ...
}
```

File: src/main/java/io/appium/java_client/android/Activity.java [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+    public String getAppActivity() {
+        return appActivity;
+    }
+
+    /**
+     * Sets the app activity value.
+     *
+     * @param appActivity The app activity value.
+     */
+    public void setAppActivity(String appActivity) {
```
Comment (TikhomirovSergey) [inline src/main/java/io/appium/java_client/android/Activity.java]: I'm not sure that it is necessary to keep this method. Maybe it is more senseful to keep only getter for activity and make this field final. @SrinivasanTarget What is your opinion?

File: src/main/java/io/appium/java_client/android/Activity.java [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+    public String getAppPackage() {
+        return appPackage;
+    }
+
+    /**
+     * Sets the app package value.
+     *
+     * @param appPackage The app package value.
+     */
+    public void setAppPackage(String appPackage) {
```
Comment (TikhomirovSergey) [inline src/main/java/io/appium/java_client/android/Activity.java]: I'm not sure that it is necessary to keep this method. Maybe it is more senseful to keep only getter for package and make this field final. @SrinivasanTarget What is your opinion?

**Comment by TikhomirovSergey** [comment]
@email2vimalraj I have finished the reviewing. It is almost ok. @email2vimalraj Could you read comments and make some improvenents?

ping @SrinivasanTarget

File: src/main/java/io/appium/java_client/android/Activity.java (hunk 7 above)
Comment (SrinivasanTarget) [inline src/main/java/io/appium/java_client/android/Activity.java]: +1 - Making this final would be a better option 👍

File: src/main/java/io/appium/java_client/android/Activity.java (hunk 5 above)
Comment (SrinivasanTarget) [inline src/main/java/io/appium/java_client/android/Activity.java]: Good catch 👍

**Comment by email2vimalraj** [comment]
@TikhomirovSergey and @SrinivasanTarget : Done the changes as recommended

**Comment by TikhomirovSergey** [comment]
@email2vimalraj I have approved these changes. @SrinivasanTarget ?

**Comment by SrinivasanTarget** [comment]
@email2vimalraj It's a merge away. Waiting for Codacy to be happy.

**Comment by SrinivasanTarget** [comment]
Thanks @email2vimalraj  👍

**Pull request merged**